### PR TITLE
Push back excessive requests for stats

### LINKED ISSUE #51992: Push back on excessive requests for stats
Today stats requests are processed by the `management` threadpool, which is also used for important internal management tasks such as syncing global checkpoints and retention leases, ILM and SLM. The `management` threadpool has an unbounded queue. Some stats take a nontrivial amount of effort to compute and it is certainly possible to request stats more frequently than the cluster can respond. We cannot control the behaviour of clients requesting stats, and I've seen more than a few situations where an errant monitoring system harms the cluster with its request rate (see links in #51915). Since we doggedly enqueue every request it can take a very long time to recover from this situation, and while working through the queue the well-behaved internal management tasks do not run in a timely fashion. The quickest recovery path may be to restart any affected nodes.

I think we should be pushing back against this kind of behaviour to protect the cluster from abusive monitoring clients. We could, for instance, use different threadpools for the internal (and well-behaved) actions from external (and possibly-abusive) ones, and use a bounded queue for the threadpool handling the external actions. Some users of the management threadpool are not clearly one or the other and we'll need to use some judgement to decide whether we need to protect them from abuse - e.g. license management, security cache management. I've yet to see such actions involved in struggling clusters, however, so perhaps either way would be ok.

Relates #51915

### PR DESCRIPTION
**Pull request by gmarouli**
### The issue we are addressing
We are trying to avoid overwhelming the coordinating node of a stats or a recovery request in the case where other nodes are having troubles. See #51992 :
 
> if a single node becomes unresponsive for a few minutes then we could see a couple hundred requests build up and in a decent sized cluster each could consume many MBs of heap on the coordinating node.

### Approach
We are introducing a limit on how many stats and recovery requests a node can coordinate concurrently. When we receive one too many of this requests the node rejects with a `409`. We are not concerned with further handling of this error because we assume that these requests would have timed out anyway. The limit is configurable as a cluster setting. 

The choice to have one limit that would apply to both requests was made for the sake of simplicity. This is seen as a protective mechanism that is triggered only to defend the node if the cluster is having troubles and not to rate limit requests when things are going well. We could alternatively create separate limits for each requests we want to bound, if there are objections to this approach. 

### Implementation details
We introduce `StatsRequestLimiter` which uses `AdjustableSemaphore` to bound the requests. Via the method `tryToExecute(Task task, Request request, ActionListener<Response> listener, TriConsumer<Task, Request, ActionListener<Response>> execute)`, this class orchestrates all aspects of pushing back, it checks the semaphore, it tracks metrics and handles exceptions.

### Affected actions:
- `IndicesStatsActions`
- `RecoveryAction`
- `IndicesSegmentsAction` 
-  `NodesStatsAction`
- `ClusterStatsAction`
- `NodesInfoAction`
- `NodesUsageAction`

Resolves #51992

### CODE REVIEW

File: docs/reference/modules/cluster/misc.asciidoc
Comment (dakrone) [inline docs/reference/modules/cluster/misc.asciidoc:119]: It would be good to include what the current default is here also, I think

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:30]: It doesn't appear that we use this integer anywhere, outside of when we're creating or modifying the semaphore. Can we remove it since the integer value is unneeded?

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:23]: Out of curiosity, how did we arrive at 20 for the default value of this?

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks for working on this Mary, I left a few small comments.

Another question I have is, is this something that it makes sense to track as a stats metric? It seems like it might be useful to have a "number of times diagnostics were requested but rejected" similar to the way that we track threadpool rejections. Having something like that would be useful for an administrator/support to be able to tweak the setting to help when stats requests are overloading a cluster. 

If that sounds useful maybe we can think about it as a follow-up to this?

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:23]: That is kind of a guess, I was planning to check with you if you have a better estimation. My line of thought:

I assume that it's common to poll every second and that 10 seconds is a long time to wait for the request. Based on that to hit the limit of the calls if the limit was 10, you would need 10 requests to hit the same node and last more than 10 seconds. Then I doubled it just to be sure that this will only be triggered if there is real trouble. 

What do you think?

File: docs/reference/modules/cluster/misc.asciidoc
Comment (gmarouli) [reply docs/reference/modules/cluster/misc.asciidoc:119]: Definitely, I will add it.

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:30]: You are right, I used it during my development to double check that it worked as expected, but it doesn't add anything in the long run.... I will remove it.

**Comment by gmarouli** [comment]
> Another question I have is, is this something that it makes sense to track as a stats metric? It seems like it might be useful to have a "number of times diagnostics were requested but rejected" similar to the way that we track threadpool rejections. Having something like that would be useful for an administrator/support to be able to tweak the setting to help when stats requests are overloading a cluster.

This is a very good point. I think it should be part of this PR if it's not exploding the scope of it because it makes it a more rounded feature. If we want this to provide enough information to the users then I would say we need more than the rejected requests right? Shouldn't we monitor all three aspects:
- Acquiring
- Releasing
- Rejecting

The reason that I am suggesting this, is maybe the problem is apparent when the requests are rejected but there are usually earlier signs that could be useful and can help users find the beginning of the problem. What do you think??

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:23]: The number I had in mind was more like 100.

The typical polling interval with things like Metricbeat is more like 10s, but it hits multiple endpoints in each cycle. Not sure exactly how many, but I expect it's ≤10. In addition there's going to be ad-hoc stats requests (e.g. running the support diagnostic tool). 20 feels a little tight to me. The point of this feature is as a last-ditch protection against runaway stats requests leading to an OOM, just in case we have a broken client that doesn't cancel on a timeout, or measures like #82685 etc aren't quite enough. In the cases I've investigated the node has usually managed to collect several hundred sets of responses before falling over.

Obviously the OOM point is a function of the heap size of the node, and the size of the responses, and therefore the type of request and the size and complexity of the cluster. Ideally we'd impose an ergonomic limit based on these factors, but in practice I don't think it's worth the effort: a flat 100 would fix all the cases I've seen whilst not limiting reasonable usage.

**Review by DaveCTurner — COMMENTED** [review]
I'm undecided whether we should exclude requests that specify target indices. Monitoring clients just request everything, and these are the ones that cause trouble when they build up, but a client that just asks for stats about a single index will cause fewer problems; moreover such a client might want to make many more requests in parallel.

There's a few `TransportNodesAction` subclasses that we might also want to consider limiting, particularly cluster stats and node stats.

**Comment by DaveCTurner** [comment]
>  Shouldn't we monitor all three aspects:

In threadpool stats we track counts of completed and rejected actions, cumulative over the life of each node, and also report the number of currently-active threads (and various other things that don't translate to this situation). I find the currently-active value to be particularly useful when visualised as a time series. So +1 to tracking these things here too.

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:22]: Naming suggestion:

```suggestion
        "node.stats.max_concurrent_requests",
```

`max` and `bounded` are redundant, we mostly call them `stats` not `diagnostics`, and I think we can put this under `node.*` rather than calling it `cluster..._per_node`.

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java:33]: Naming suggestion:

```suggestion
public abstract class TransportBroadcastByNodeStatsAction<
```

The fact that it's bounded is something of an implementation detail, we may want to apply other behaviour across all such stats actions in future too (e.g. tracking how many of them are completed over time)

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:19]: Naming suggestion

```suggestion
public class StatsRequestLimiter {
```

**Review by DaveCTurner — COMMENTED** [review]
Sorry for the quick-fire reviews, this one is just alternative naming suggestions.

**Comment by gmarouli** [comment]
> I'm undecided whether we should exclude requests that specify target indices. Monitoring clients just request everything, and these are the ones that cause trouble when they build up, but a client that just asks for stats about a single index will cause fewer problems; moreover such a client might want to make many more requests in parallel.

Hm.... I am not sure if this is worth the complexity. I am concerned that it makes it more difficult to explain to the users and it is not as easy to know where to draw the line? What if the index expression has a wildcard that matches almost all indices? So, I would like to add this only if it really adds value, for example what's the worst thing that can happen if we do not distinguish between the two? What I can think about is that a user does a separate call for every single index they have and that exceeds the limit, but is that a common case?


> There's a few `TransportNodesAction` subclasses that we might also want to consider limiting, particularly cluster stats and node stats.

I think it makes sense to add them too, if they could cause the same issue, now that hopefully the stats calls will not it makes sense to add them. That would change the structure of the code a bit but I think it's worth it.

I will snoop around the code to try to come up with a list of relevant cases, but if you have things in mind already like you said cluster and node stats feel free to add them :).

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java:22]: > Sorry for the quick-fire reviews, this one is just alternative naming suggestions.

Thank you, this helps a lot to get the lingo right :)

**Comment by gmarouli** [comment]
Suggested calls to be limited:
- `TransportIndicesSegmentsAction` extends `TransportBroadcastByNodeAction`
- `TransportNodesStatsAction` extends `TransportNodesAction`
- `TransportClusterStatsAction` extends `TransportNodesAction`
- `TransportNodesHotThreadsAction` extends `TransportNodesAction` (I am not sure if this is too specialized so it should not be limited)
- `TransportNodesInfoAction` extends `TransportNodesAction`
- `TransportNodesUsageAction` extends `TransportNodesAction`

Internal actions are excluded and more specialized stats such as `GeoIpDownloaderStatsTransportAction`. What do you think about this list?

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:84]: I think this is missing a word?
```suggestion
            listener.onFailure(new EsRejectedExecutionException("too many concurrent stats requests"));
```

Also, perhaps we can mention the limit in the error message, something like "too many concurrent stats requests (limit: 100)"?

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:81]: I *think* (but I'm not sure) this can double-invoke the `release.run()`.

Since the request wraps the `release::run` *before* the original listener, the `executeAction` can run, the first invocation of `release.run()` can happen (the one from `runBefore`), then the original listener can throw an exception, which would cause the `success` boolean not to be updated, and then the `release.run()` in the `finally` block would run a second time.

I guess it depends on whether ActionListeners are expected to ever throw exceptions (@DaveCTurner can probably clarify on this point), but if they can, then maybe the `runBefore` call could do something like `() -> { success.getAndSet(true); release.run(); }` and then it wouldn't have the potential for invoking twice?

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:94]: I think this can be:
```suggestion
        return new StatsRequestStats(List.copyOf(stats.values()));
```

File: server/src/test/java/org/elasticsearch/action/support/broadcast/node/AbstractTransportBroadcastByNodeActionTestCase.java
Comment (dakrone) [inline server/src/test/java/org/elasticsearch/action/support/broadcast/node/AbstractTransportBroadcastByNodeActionTestCase.java:56]: This abstract class is never used by anything in this PR, is it a leftover?

**Review by dakrone — COMMENTED** [review]
This generally looks good to me, I left a couple more comments, but I'd also like to wait for David's review also (and confirmation about the double-release comment) before merging

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java]: I wasn't careful enough, this would make a list of `StatsHolder` but we need the actual stats

File: server/src/test/java/org/elasticsearch/action/support/broadcast/node/AbstractTransportBroadcastByNodeActionTestCase.java
Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/action/support/broadcast/node/AbstractTransportBroadcastByNodeActionTestCase.java:56]: Good catch! It's a leftover

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:84]: That could be helpful. I will add it.

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:81]: I am processing the possibilities but I do not have an answer yet. Coming asap

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java]: I converted it to a different onliner though.

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java]: I changed the structure of the code to expand to more stats requests, I chose to create a wrapper that the stats actions could use, in this set-up this naming isn't applicable.

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:80]: Yes I believe it's possible that `executeAction` might throw an exception but still go on to complete its listener. I mean it probably shouldn't, but in general we're ok with completing an `ActionListener` multiple times so we don't protect against this.

However here the `release` runnable is a `RunOnce` so it doesn't matter.

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java]: Oh, hmm, looks like this was pending from a much older review. Still stands tho I think.

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:80]: I will start with the suggestion that the `runBefore()` does something like this `() -> { success.getAndSet(true); release.run(); }` because I feel more confident about my analysis :). I believe this will cause double invocation in almost all cases, because:

1. `executeAction` introduces the `runBefore` to run before the original action listener starts whatever async needs to be run and finishes
2. Chances are by this point, the async work hasn't finished yet, which means the `success` is false and the `finally` will run the `release.run()` effectively releasing the semaphore while the node is still coordinating
3. When the async work finishes, the original listener will be notified and the `release.run()` will be called again.

Now let's go to the trickier part, it's not clear to me yet where could the original listener throw an exception. If I look at the code of the `RunBeforeActionListener`:
```
        @Override
        public void onResponse(T response) {
            try {
                runBefore.run();
            } catch (Exception ex) {
                super.onFailure(ex);
                return;
            }
            delegate.onResponse(response);
        }

```

Similar the failure is being handled,  the `release` will be called even if the original listener has an error. I would expect that the only situation where this can go wrong, is if the listener is never completed and the `executeAction` finished. But this effectively would mean that the node is still coordinating so we have a different problem then right?

Did I address your concern properly? Or am I still missing something?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:80]: I think there's no need for any changes in this area - `RunOnce` already does what is needed.

It's not just about completing the listener once but also releasing things in the `finally` block, you also need to protect against completing the listener multiple times. But we already do the right thing here.

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBoundedDiagnosticAction.java]: On it.

File: server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java:80]: Yep, totally didn't notice that the `release` was a `RunOnce`, that's fine then and no worries about this.

**Review by dakrone — APPROVED** [review]
LGTM, thanks for iterating Mary!

**Comment by gmarouli** [comment]
Thanks for the review @dakrone & @DaveCTurner . With your input the MR is much better than the initial effort!

**Comment by joegallo** [comment]
I reverted this from the 8.2 branch via #85504, and I've updated the version tag to here to v8.3.0 rather than v8.2.0.

**Pull request merged**
